src/control/cmd/daos_server/storage_scm.go

```diff
@@ -35,11 +41,61 @@ type scmCmd struct {
 type prepareSCMCmd struct {
 	cmdutil.LogCmd `json:"-"`
 	helperLogCmd   `json:"-"`
+	cfgCmd         `json:"-"`
+	socketCmd      `json:"-"`
 
 	NrNamespacesPerSocket uint `short:"S" long:"scm-ns-per-socket" description:"Number of PMem namespaces to create per socket" default:"1"`
 	Force                 bool `short:"f" long:"force" description:"Perform SCM operations without waiting for confirmation"`
 }
 
+func setSockFromCfg(log logging.Logger, cfg *config.Server, affSrc config.EngineAffinityFn, req *storage.ScmPrepareRequest) error {
+	msgSkip := "so skip fetching sockid from config"
+
+	if cfg == nil {
+		log.Debug("nil input server config " + msgSkip)
+		return nil
+	}
+	if req.SocketID != nil {
+		log.Debug("sockid set in req " + msgSkip)
+		return nil
+	}
+
+	// Set engine NUMA affinities accurately in config.
+	if err := cfg.SetEngineAffinities(log, affSrc); err != nil {
+		log.Error(fmt.Sprintf("failed to set engine affinities %s: %s", msgSkip, err))
+		return nil
+	}
+
+	socksToPrep := make(map[uint]bool)
+
+	for _, ec := range cfg.Engines {
+		for _, scmCfg := range ec.Storage.Tiers.ScmConfigs() {
+			if scmCfg.Class == storage.ClassRam {
+				continue
+			}
+			socksToPrep[ec.Storage.NumaNodeIndex] = true
+		}
+	}
+
+	// Only set socket ID in request if one engine has been configured in config file with SCM
+	// storage specified, if engines exist assigned to different sockets, don't set the value.
+
+	switch len(socksToPrep) {
+	case 0:
+	case 1:
+		log.Debug("config contains scm-engines attached to a single socket, " +
+			"only process that socket")
+		for k := range socksToPrep {
+			req.SocketID = &k
+		}
+	default:
```

Comment:
Why should we prepare all sockets if we have 4 sockets and only 2 used by an engine ?

Reply:
This is indeed a deficiency in the implementation which I'm hoping to address at some point. The reason this is not trivial to address is that the scm backend doesn't take a list of socket IDs and will only process 1 or all sockets. Solutions include updating the backend to take more than one socket ID or calling the backend multiple times from the above code. This task needs to be scheduled and prioritised when compared to other work. At the moment the workaround would be just to run the prepare command twice once for each socket in use.

Reply:
OK